### DISCUSSION TITLE
Tracing recombination events

### DISCUSSION THREAD

**Discussion by jikhashkya** [tskit-dev/msprime#1868] [General]
What's the best way to track the time of the recombination events that gave rise to IBD segments' boundaries ? I followed [this](https://github.com/tskit-dev/tutorials/issues/43) link for some insight and it was very helpful but I'm not entirely sure if the recombination events pertaining to all IBD segments' boundaries can be tracked successfully ? Any help would be much appreciated.

**Comment by jeromekelleher** [comment]
I think the best option is to use the [record_full_arg](https://tskit.dev/msprime/docs/latest/ancestry.html#ancestral-recombination-graph) option. This will stored nodes associated with each RE event, so you can trace back the source of each breakpoint to a particular event. I can help a bit more if you're having trouble pulling out this information.

**Reply by jikhashkya** [reply]
Hi there again @jeromekelleher , so I tried the following code snippet to first simulate the tree sequence with ARG.

```
import msprime
import tskit

num_samples = 4
ts = msprime.sim_ancestry(num_samples, sequence_length=10,
                          recombination_rate=0.1, random_seed=3,
                          record_full_arg=True, ploidy=2)
version = 1
with open("pic-{}.txt".format(version), 'w') as out:
    out.write(ts.draw_text())

with open('nodes.info-{}'.format(version), 'w') as nodes:
    ts.dump_text(nodes=nodes)
```
Following are the tree sequence  and nodes table that I obtain: 

```
3.48┊          32     ┊                 ┊                 ┊                 ┊                 ┊
    ┊        ┏━━┻━━┓  ┊                 ┊                 ┊                 ┊                 ┊
2.46┊        ┃    31  ┊                 ┊                 ┊        31       ┊          31     ┊
    ┊        ┃     ┃  ┊                 ┊                 ┊     ┏━━━┻━━━┓   ┊       ┏━━━┻━━┓  ┊
2.20┊       29     ┃  ┊                 ┊                 ┊     ┃      30   ┊      30      ┃  ┊
    ┊        ┃     ┃  ┊                 ┊                 ┊     ┃       ┃   ┊       ┃      ┃  ┊
2.19┊        ┃    28  ┊                 ┊                 ┊    28       ┃   ┊       ┃     28  ┊
    ┊        ┃     ┃  ┊                 ┊                 ┊     ┃       ┃   ┊       ┃      ┃  ┊
1.86┊       27     ┃  ┊        27       ┊                 ┊     ┃      27   ┊      27      ┃  ┊
    ┊     ┏━━┻━━┓  ┃  ┊     ┏━━━┻━━━┓   ┊                 ┊     ┃       ┃   ┊       ┃      ┃  ┊
1.74┊    25     ┃  ┃  ┊    25       ┃   ┊                 ┊    26       ┃   ┊       ┃     26  ┊
    ┊     ┃     ┃  ┃  ┊     ┃       ┃   ┊                 ┊     ┃       ┃   ┊       ┃      ┃  ┊
1.35┊     ┃    24  ┃  ┊     ┃      24   ┊          24     ┊     ┃      24   ┊      24      ┃  ┊
    ┊     ┃     ┃  ┃  ┊     ┃     ┏━┻┓  ┊        ┏━━┻━━┓  ┊     ┃     ┏━┻┓  ┊     ┏━┻━━┓   ┃  ┊
1.24┊     ┃     ┃ 22  ┊     ┃     ┃ 23  ┊        ┃    23  ┊     ┃     ┃ 23  ┊     ┃   23   ┃  ┊
    ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊        ┃     ┃  ┊     ┃     ┃  ┃  ┊     ┃    ┃   ┃  ┊
1.10┊     ┃    21  ┃  ┊     ┃    21  ┃  ┊       21     ┃  ┊     ┃    21  ┃  ┊    21    ┃   ┃  ┊
    ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┏━━┻━━┓  ┃  ┊     ┃     ┃  ┃  ┊     ┃    ┃   ┃  ┊
0.93┊    20     ┃  ┃  ┊    20     ┃  ┃  ┊     ┃     ┃  ┃  ┊    20     ┃  ┃  ┊     ┃    ┃  20  ┊
    ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃    ┃   ┃  ┊
0.76┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊    18     ┃  ┃  ┊    19     ┃  ┃  ┊     ┃    ┃  19  ┊
    ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃    ┃   ┃  ┊
0.42┊     ┃    17  ┃  ┊     ┃    17  ┃  ┊     ┃    17  ┃  ┊     ┃    17  ┃  ┊    17    ┃   ┃  ┊
    ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊   ┏━┻━┓  ┃   ┃  ┊
0.33┊    15     ┃  ┃  ┊    15     ┃  ┃  ┊    16     ┃  ┃  ┊    16     ┃  ┃  ┊   ┃   ┃  ┃  16  ┊
    ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊     ┃     ┃  ┃  ┊   ┃   ┃  ┃   ┃  ┊
0.12┊    14     ┃  ┃  ┊    14     ┃  ┃  ┊    14     ┃  ┃  ┊    14     ┃  ┃  ┊   ┃   ┃  ┃  14  ┊
    ┊   ┏━┻━━┓  ┃  ┃  ┊   ┏━┻━━┓  ┃  ┃  ┊   ┏━┻━━┓  ┃  ┃  ┊   ┏━┻━━┓  ┃  ┃  ┊   ┃   ┃  ┃   ┃  ┊
0.12┊  12    ┃  ┃  ┃  ┊  12    ┃  ┃  ┃  ┊  12    ┃  ┃  ┃  ┊  12    ┃  ┃  ┃  ┊  13   ┃  ┃   ┃  ┊
    ┊   ┃    ┃  ┃  ┃  ┊   ┃    ┃  ┃  ┃  ┊   ┃    ┃  ┃  ┃  ┊   ┃    ┃  ┃  ┃  ┊   ┃   ┃  ┃   ┃  ┊
0.10┊   ┃   11  ┃  ┃  ┊   ┃   11  ┃  ┃  ┊   ┃   11  ┃  ┃  ┊   ┃   11  ┃  ┃  ┊   ┃   ┃  ┃  11  ┊
    ┊   ┃   ┏┻┓ ┃  ┃  ┊   ┃   ┏┻┓ ┃  ┃  ┊   ┃   ┏┻┓ ┃  ┃  ┊   ┃   ┏┻┓ ┃  ┃  ┊   ┃   ┃  ┃  ┏┻┓ ┊
0.09┊  10   ┃ ┃ ┃  ┃  ┊  10   ┃ ┃ ┃  ┃  ┊  10   ┃ ┃ ┃  ┃  ┊  10   ┃ ┃ ┃  ┃  ┊  10   ┃  ┃  ┃ ┃ ┊
    ┊ ┏━┻┓  ┃ ┃ ┃  ┃  ┊ ┏━┻┓  ┃ ┃ ┃  ┃  ┊ ┏━┻┓  ┃ ┃ ┃  ┃  ┊ ┏━┻┓  ┃ ┃ ┃  ┃  ┊ ┏━┻┓  ┃  ┃  ┃ ┃ ┊
0.02┊ ┃  ┃  ┃ ┃ ┃  9  ┊ ┃  ┃  ┃ ┃ ┃  9  ┊ ┃  ┃  ┃ ┃ ┃  9  ┊ ┃  ┃  ┃ ┃ ┃  9  ┊ ┃  ┃  ┃  9  ┃ ┃ ┊
    ┊ ┃  ┃  ┃ ┃ ┃ ┏┻┓ ┊ ┃  ┃  ┃ ┃ ┃ ┏┻┓ ┊ ┃  ┃  ┃ ┃ ┃ ┏┻┓ ┊ ┃  ┃  ┃ ┃ ┃ ┏┻┓ ┊ ┃  ┃  ┃ ┏┻┓ ┃ ┃ ┊
0.01┊ ┃  8  ┃ ┃ ┃ ┃ ┃ ┊ ┃  8  ┃ ┃ ┃ ┃ ┃ ┊ ┃  8  ┃ ┃ ┃ ┃ ┃ ┊ ┃  8  ┃ ┃ ┃ ┃ ┃ ┊ ┃  8  ┃ ┃ ┃ ┃ ┃ ┊
    ┊ ┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃ ┊ ┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃ ┊ ┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃ ┊ ┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃ ┊ ┃ ┏┻┓ ┃ ┃ ┃ ┃ ┃ ┊
0.00┊ 0 1 2 6 7 3 4 5 ┊ 0 1 2 6 7 3 4 5 ┊ 0 1 2 6 7 3 4 5 ┊ 0 1 2 6 7 3 4 5 ┊ 0 1 2 3 4 5 6 7 ┊
    0                 3                 5                 6                 7                10
```


```
╔══╤══════╤══════════╤══════════╤══════════╤════════╗
║id│flags │population│individual│time      │metadata║
╠══╪══════╪══════════╪══════════╪══════════╪════════╣
║0 │     1│         0│         0│0.00000000│     b''║
║1 │     1│         0│         0│0.00000000│     b''║
║2 │     1│         0│         1│0.00000000│     b''║
║3 │     1│         0│         1│0.00000000│     b''║
║4 │     1│         0│         2│0.00000000│     b''║
║5 │     1│         0│         2│0.00000000│     b''║
║6 │     1│         0│         3│0.00000000│     b''║
║7 │     1│         0│         3│0.00000000│     b''║
║8 │     0│         0│        -1│0.00523932│     b''║
║9 │     0│         0│        -1│0.01755782│     b''║
║10│     0│         0│        -1│0.09416924│     b''║
║11│     0│         0│        -1│0.10246509│     b''║
║12│131072│         0│        -1│0.11714256│     b''║
║13│131072│         0│        -1│0.11714256│     b''║
║14│     0│         0│        -1│0.12320888│     b''║
║15│131072│         0│        -1│0.33135423│     b''║
║16│131072│         0│        -1│0.33135423│     b''║
║17│     0│         0│        -1│0.41871600│     b''║
║18│131072│         0│        -1│0.76244860│     b''║
║19│131072│         0│        -1│0.76244860│     b''║
║20│262144│         0│        -1│0.92612452│     b''║
║21│     0│         0│        -1│1.10489666│     b''║
║22│131072│         0│        -1│1.24037929│     b''║
║23│131072│         0│        -1│1.24037929│     b''║
║24│     0│         0│        -1│1.34753710│     b''║
║25│131072│         0│        -1│1.74088048│     b''║
║26│131072│         0│        -1│1.74088048│     b''║
║27│     0│         0│        -1│1.86281624│     b''║
║28│262144│         0│        -1│2.19013359│     b''║
║29│131072│         0│        -1│2.20170578│     b''║
║30│131072│         0│        -1│2.20170578│     b''║
║31│     0│         0│        -1│2.46403664│     b''║
║32│     0│         0│        -1│3.48232723│     b''║
╚══╧══════╧══════════╧══════════╧══════════╧════════╝
```

From what I understand the recombination nodes should have the flag **NODE_IS_RE_EVENT** in the nodes table signified by the identifier(131072) which we see in the above table. The breakpoints and recombination events (from what I understand are as follows):

| RE_EVENT | Breakpoint |
|-------------|-------------|
|       12/13   |        7         |
|       15/16   |        5         |
|       18/19   |        6         |
|       22/23   |        3         |
|       25/26   |        5?6?    |
|       29/30   |        3?6?    |

Now, this might be rare case of multiple recombination events at the same breakpoint, but is there anyway to distinguish it if that occurs ? And, what would be a good way to relate the recombination events with the breakpoints ( as well as IBDs)?

**Reply by jeromekelleher** [reply]
Good question @jikhashkya, you're right it's not immediately straightforward to pull this information out. Here's what I've come up with:

```python
tables = ts.tables
re_nodes = np.where(tables.nodes.flags == msprime.NODE_IS_RE_EVENT)[0]
for node_a, node_b in re_nodes.reshape((-1, 2)):
    event_time = tables.nodes.time[node_a]
    print(f"RE: {node_a} {node_b} @ t={event_time:.2g}", end=" ")
    assert tables.nodes.time[node_b] == event_time
    a_edges = tables.edges[tables.edges.parent == node_a]
    b_edges = tables.edges[tables.edges.parent == node_b]
    a_right = np.max(a_edges.right)
    b_left = np.min(b_edges.left)
    if a_right == b_left:
        print("breakpoint at", a_right)
    else:
        print(f"breakpoint in non ancestral between {a_right} and {b_left}")
```

which gives

```
RE: 12 13 @ t=0.12 breakpoint at 7.0
RE: 15 16 @ t=0.33 breakpoint at 5.0
RE: 18 19 @ t=0.76 breakpoint at 6.0
RE: 22 23 @ t=1.2 breakpoint at 3.0
RE: 25 26 @ t=1.7 breakpoint in non ancestral between 5.0 and 6.0
RE: 29 30 @ t=2.2 breakpoint in non ancestral between 3.0 and 6.0
```

So, we look at the RE nodes pair-by-pair and then examine the corresponding edges. If the edges for the two nodes are adjacent then the RE event happened in ancestral material and we can pinpoint it. If not, we only that it happened somewhere in non ancestral material (but that it doesn't contribute any breakpoints).

@JereKoskela - do you agree with my analysis here?

**Reply by JereKoskela** [reply]
Yes, that looks right to me @jeromekelleher. And since your code iterates over recombination events rather than just break points, the output will tell you if a recombination happened at the same point more than once.

**Reply by jikhashkya** [reply]
@jeromekelleher  - Thank you for the help. As for the second part of my question, let's take for example **sample pair 0 and 4** and the IBD segments shared by these two samples specified by the genomic position intervals (left , right) are:

1. (0, 3) 
2. (3, 5) 
3. (5, 6) 
4. (6, 10)

Now, trying to relate the RE event that led to these IBD segments, would it be correct to say that (0,3) IBD segment was the result of RE: 22/23, (3,5) was the result of RE:22/23 on the left side and RE:15/16 on the right and similarly, for the rest of the IBD segments for this sample pair ? Additionally, for **sample pair 0 and 3**, one IBD segment covers the positions (0,5) _(with MRCA 27)_ , so is RE:15/16 also responsible for this IBD segment ?

**Reply by jeromekelleher** [reply]
This isn't obvious to me I'm afraid @jikhashkya - is there always going to be a one-to-one correspondence between IBD segments and recombination events? I guess we can trace back the source of a given breakpoint to an event (or events), and you could say that these "cause" the specific IBD segments?

**Reply by jikhashkya** [reply]
I apologize if the question is getting away from msprime-specific discussions, but from my understanding the boundaries of an IBD segment are results of some ancient recombination events, but tracing back to those events might not be as trivial ?

**Reply by jeromekelleher** [reply]
I think it's tricky, because we have breakpoints in non-ancestral material they may also give rise to IBD segment breaks which align with breakpoints that were actually caused by recombinations in ancestral material. It's not obvious to me which event cause the breakpoint then, or how you'd disambiguate this situation (or, whether it actually matters at all!). I'm sure there's a way to do it, but I'd have to spend a good bit of time on it to be confident that it's the right answer.

Sorry I can't be more helpful!

**Reply by jikhashkya** [reply]
I appreciate  your help anyways. Thank you @jeromekelleher
